docs/typescript/examples/checkout-tutorial.md

```diff
@@ -1,299 +1,370 @@
+      });
+    if (numAffected <= 0) {
+      throw new Error('Insufficient Inventory');
+    }
   }
 
-  return generatePaymentUrls(handle.workflowID, session_id);
-}
-```
-
-## The Checkout Workflow
+  @DBOS.transaction()
+  static async undoSubtractInventory(): Promise<void> {
+    await DBOS.knexClient<Product>('products')
+      .where({ product_id: PRODUCT_ID })
+      .update({ inventory: DBOS.knexClient.raw('inventory + ?', 1) });
+  }
 
-The checkout workflow reserves inventory for an order, attempts to process payment, and fufills the order if payment is successful.
-As we'll show, it's **reliable**: it always fulfills orders if payments succeed, never charges customers twice for the same order, and always returns reserved inventory on failure.
+  @DBOS.transaction()
+  static async setInventory(inventory: number): Promise<void> {
+    await DBOS.knexClient<Product>('products').where({ product_id: PRODUCT_ID }).update({ inventory });
+  }
 
-:::info
-Check out our [e-commerce demo app](https://github.com/dbos-inc/dbos-demo-apps/tree/main/typescript/e-commerce) for a more elaborate example.
-:::
+  @DBOS.transaction({ readOnly: true })
+  static async retrieveProduct(): Promise<Product> {
+    const item = await DBOS.knexClient<Product>('products').select('*').where({ product_id: PRODUCT_ID });
+    if (!item.length) {
+      throw new Error(`Product ${PRODUCT_ID} not found`);
+    }
+    return item[0];
+  }
 
-### Registering the workflow
+  @DBOS.transaction()
+  static async createOrder(): Promise<number> {
+    const orders = await DBOS.knexClient<Order>('orders')
+      .insert({
+        order_status: OrderStatus.PENDING,
+        product_id: PRODUCT_ID,
+        last_update_time: DBOS.knexClient.fn.now(),
+        progress_remaining: 10,
+      })
+      .returning('order_id');
+    const orderID = orders[0].order_id;
+    return orderID;
+  }
 
-First, we declare the workflow using the `@DBOS.workflow` decorator:
+  @DBOS.transaction()
+  static async markOrderPaid(order_id: number): Promise<void> {
+    await DBOS.knexClient<Order>('orders').where({ order_id: order_id }).update({
+      order_status: OrderStatus.PAID,
+      last_update_time: DBOS.knexClient.fn.now(),
+    });
+  }
 
-```javascript
-@DBOS.workflow()
-static async checkoutWorkflow(): Promise<void> {
-```
+  @DBOS.transaction()
+  static async errorOrder(order_id: number): Promise<void> {
+    await DBOS.knexClient<Order>('orders').where({ order_id: order_id }).update({
+      order_status: OrderStatus.CANCELLED,
+      last_update_time: DBOS.knexClient.fn.now(),
+    });
+  }
 
-### Reserving inventory
-Before purchasing an item, the checkout workflow reserves inventory for the order using the `reserveInventory` transaction.
-If this fails (likely because the item is out of stock), the workflow notifies its handler of the failure using the [events API](../tutorials/workflow-tutorial.md#workflow-events) and returns.
+  @DBOS.transaction({ readOnly: true })
```

Comment:
We can update the demo app in a separate PR. But we can just remove readOnly here.
```suggestion
  @DBOS.transaction()
```